src/Evaluator.hs

```diff
@@ -263,28 +268,47 @@ withScopeOf scope expr = do
     Syntax (Stx scopeSet loc _) ->
       pure $ ValueSyntax $ Syntax $ Stx scopeSet loc expr
 
-doCase :: Value -> [(Pattern, Core)] -> Eval Value
+doDataCase :: Value -> [(ConstructorPattern, Core)] -> Eval Value
+doDataCase v0 [] = throwError (EvalErrorCase v0)
```

Comment:
Having an empty list of cases seems more like a compile-time error than a run-time error, we could use a `NonEmpty` in `CoreDataCase` to ensure that we catch it earlier.

Reply:
This should be dynamic, for two reasons:
 1. It's the base case for recursion when no pattern matches (we have no exhaustiveness checking ATM)
 2. Empty cases are the natural eliminator for constructorless datatypes, witnessing unreachability.